src/coreclr/jit/optimizer.cpp

```diff
@@ -6907,52 +6908,79 @@ bool Compiler::optHoistThisLoop(unsigned lnum, LoopHoistContext* hoistCtxt, Basi
     // Until we have post-dominators, we'll special-case for single-exit blocks.
     //
     // Todo: it is not clear if this is a correctness requirement or a profitability heuristic.
-    // It seems like the latter. Ideally have enough safeguards to prevent hoisting exception
-    // or side-effect dependent things.
+    // It seems like the latter. Ideally there are enough safeguards to prevent hoisting exception
+    // or side-effect dependent things. Note that HoistVisitor uses `m_beforeSideEffect` to determine if it's
+    // ok to hoist a side-effect. It allows this only for the first block (the entry block), before any
+    // side-effect has been seen. After the first block, it assumes that there has been a side effect and
+    // no further side-effect can be hoisted. It is true that we don't analyze any program behavior in the
+    // flow graph between the entry block and the subsequent blocks, whether they be the next block dominating
+    // the exit block, or the pre-headers of nested loops.
     //
     // We really should consider hoisting from conditionally executed blocks, if they are frequently executed
     // and it is safe to evaluate the tree early.
     //
     ArrayStack<BasicBlock*> defExec(getAllocatorLoopHoist());
-    BasicBlockList*         preHeadersList = existingPreHeaders;
+
+    // Add the pre-headers of any child loops to the list of blocks to consider for hoisting.
+    // Note that these are not definitely executed. However, it is a heuristic that they will
+    // often provide good opportunities for further hoisting since we hoist from inside-out,
+    // and the inner loop may have already hoisted something loop-invariant to them. If the child
+    // loop pre-header block would be added anyway (by dominating the loop exit block), we don't
+    // add it here, and let it be added naturally, below.
+    //
+    // Note that all pre-headers get added first, which means they get considered for hoisting last. It is
+    // assumed that the order does not matter for correctness (since there is no execution order known).
+    // Note that the order does matter for the hoisting profitability heuristics, as we might
+    // run out of hoisting budget when processing the blocks.
+
+    int childLoopPreHeaders = 0;
+    for (BasicBlock::loopNumber childLoop = pLoopDsc->lpChild; //
+         childLoop != BasicBlock::NOT_IN_LOOP;                 //
+         childLoop = optLoopTable[childLoop].lpSibling)
+    {
+        if (optLoopTable[childLoop].lpIsRemoved())
+        {
+            continue;
+        }
+        INDEBUG(optLoopTable[childLoop].lpValidatePreHeader());
+        BasicBlock* childPreHead = optLoopTable[childLoop].lpHead;
+        if (pLoopDsc->lpExitCnt == 1)
+        {
+            if (fgDominate(childPreHead, pLoopDsc->lpExit))
```

Comment:
So this became possible because the dominators already have the already created pre-header blocks, which was not possible earlier and hence we would just add them all in `defExec`:

https://github.com/dotnet/runtime/blob/2323fc8513c9f5e60683cf4aa90475b44a550170/src/coreclr/jit/optimizer.cpp#L6947-L6949